sirepo/auth/ldap.py

```diff
@@ -0,0 +1,102 @@
+"""LDAP authentication support
+:copyright: Copyright (c) 2022-2023 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from pykern import pkconfig
+from pykern import pkinspect
+from pykern.pkdebug import pkdlog, pkdp
+from pykern.pkcollections import PKDict
+import ldap
+import re
+import sirepo.quest
+import sirepo.util
+
+AUTH_METHOD = "ldap"
+
+AUTH_METHOD_VISIBLE = True
+
+_cfg = None
+
+# ldap dn requires escaped special chars
+_ESCAPE_DN_MAIL = re.compile(r"([,\\#<>;\"=+])")
+
+# windows login dialogue char limit for passwords
+_MAX_ENTRY = 127
+
+#: module handle
+this_module = pkinspect.this_module()
+
+#: Well known alias for auth
+user_model = "AuthEmailUser"
+
+
+class API(sirepo.quest.API):
+    @sirepo.quest.Spec("require_cookie_sentinel")
+    def api_authLdapLogin(self):
+        def _bind(creds):
+            try:
+                ldap.initialize(_cfg.server).simple_bind_s(creds.dn, creds.password)
+            except Exception as e:
+                m = "Unable to contact LDAP server"
+                if isinstance(e, ldap.INVALID_CREDENTIALS):
+                    m = "Incorrect email or password"
```

Comment:
In reading this more carefully, let's keep all the errors the same so make it a constant `_INVALID_CREDNTIALS = "Invalid user and/or password"` that's sahred between `_validate` and `_bind` so here:

```py
m = _INVALID_CREDENTIALS
```